### DISCUSSION TITLE
Setting up argo-vault-plugin for using vault as secret management in gitops

### DISCUSSION THREAD

**Discussion by dhinakar2707** [argoproj/argo-cd#12349] [Q&A]
Hi,

I have a existing argocd application running, and we are planning to use vault as secret management for the applications to store the sensitive values using argo vault plugin.

I have started the setting up the argo-vault-plugin as per the documentation, but for some how I'm not able to configure the plugin properly. below is the steps I followed for same.

1. argocd-repo-server-deploy.yaml

<img width="1213" alt="image" src="https://user-images.githubusercontent.com/45903612/217586828-074057f3-464b-4fd7-acdb-6153606ed401.png">
`apiVersion: apps/v1
kind: Deployment
metadata:
  name: argocd-repo-server
spec:
  template:
    spec:
      containers:
      - name: argocd-repo-server
        volumeMounts:
        - name: custom-tools
          mountPath: /usr/local/bin/argocd-vault-plugin
          subPath: argocd-vault-plugin

        # Note: AVP config (for the secret manager, etc) can be passed in several ways. This is just one example
        # https://argocd-vault-plugin.readthedocs.io/en/stable/config/
        envFrom:
          - secretRef:
              name: argocd-vault-plugin-credentials
      volumes:
      - name: custom-tools
        emptyDir: {}
      initContainers:
      - name: download-tools
        image: alpine:3.8
        command: [sh, -c]

        # Don't forget to update this to whatever the stable release version is
        # Note the lack of the `v` prefix unlike the git tag
        env:
          - name: AVP_VERSION
            value: "1.7.0"
        args:
          - >-
            wget -O argocd-vault-plugin
            https://github.com/argoproj-labs/argocd-vault-plugin/releases/download/v${AVP_VERSION}/argocd-vault-plugin_${AVP_VERSION}_linux_amd64 &&
            chmod +x argocd-vault-plugin &&
            mv argocd-vault-plugin /custom-tools/
        volumeMounts:
          - mountPath: /custom-tools
            name: custom-tools

      # Not strictly necessary, but required for passing AVP configuration from a secret and for using Kubernetes auth to Hashicorp Vault
      automountServiceAccountToken: `true``

2. argocd-cm.yaml
`apiVersion: v1
kind: ConfigMap
metadata:
  name: argocd-cm
data:
  configManagementPlugins: |
    - name: argocd-vault-plugin
      generate:
        command: ["argocd-vault-plugin"]
        args: ["generate", "./"]

    - name: argocd-vault-plugin-helm
      generate:
        command: ["sh", "-c"]
        args: ['helm template "$ARGOCD_APP_NAME" -n "$ARGOCD_APP_NAMESPACE" . | argocd-vault-plugin generate -']

    # This lets you pass args to the Helm invocation as described here: https://argocd-vault-plugin.readthedocs.io/en/stable/usage/#with-helm
    # IMPORTANT: passing $helm_args effectively allows users to run arbitrary code in the Argo CD repo-server.
    # Only use this when the users are completely trusted. If possible, determine which Helm arguments are needed by 
    # your users and explicitly pass only those arguments.
    - name: argocd-vault-plugin-helm-with-args
      generate:
        command: ["sh", "-c"]
        args: ['helm template "$ARGOCD_APP_NAME" -n "$ARGOCD_APP_NAMESPACE" ${helm_args} . | argocd-vault-plugin generate -']

    # This lets you pass a values file as a string as described here:
    # https://argocd-vault-plugin.readthedocs.io/en/stable/usage/#with-helm
    - name: argocd-vault-plugin-helm-with-values
      generate:
        command: ["bash", "-c"]
        args: ['helm template "$ARGOCD_APP_NAME" -n "$ARGOCD_APP_NAMESPACE" -f <(echo "$ARGOCD_ENV_HELM_VALUES") . | argocd-vault-plugin generate -']

    - name: argocd-vault-plugin-kustomize
      generate:
        command: ["sh", "-c"]
        args: ["kustomize build . | argocd-vault-plugin generate -"]`

I have used kubectl patch command to update the repo-server & configmap. 

Please can someone help me if i'm using the right approach and guide me on this.

**Comment by jaideepr97** [comment]
@ciiay any ideas here?

**Comment by ciiay** [comment]
@dhinakar2707 What error did you see? I couldn't identify the issue here.

**Reply by dhinakar2707** [reply]
@ciiay 

I have installed argocd-vault-plugin using sidecar approach. below are my codes used.

apiVersion: v1 kind: ConfigMap metadata: name: cmp-plugin data: avp-kustomize.yaml: | --- apiVersion: argoproj.io/v1alpha1 kind: ConfigManagementPlugin metadata: name: argocd-vault-plugin-kustomize spec: allowConcurrency: true # Note: this command is run _before_ anything is done, therefore the logic is to check # if this looks like a Kustomize bundle discover: find: command: - find - "." - -name - kustomization.yaml generate: command: - sh - "-c" - "kustomize build . | argocd-vault-plugin generate -" lockRepo: false avp-helm.yaml: | --- apiVersion: argoproj.io/v1alpha1 kind: ConfigManagementPlugin metadata: name: argocd-vault-plugin-helm spec: allowConcurrency: true # Note: this command is run _before_ any Helm templating is done, therefore the logic is to check # if this looks like a Helm chart discover: find: command: - sh - "-c" - "find . -name 'Chart.yaml' && find . -name 'values.yaml'" generate: # **IMPORTANT**: passing ${ARGOCD_ENV_helm_args} effectively allows users to run arbitrary code in the Argo CD # repo-server (or, if using a sidecar, in the plugin sidecar). Only use this when the users are completely trusted. If # possible, determine which Helm arguments are needed by your users and explicitly pass only those arguments. command: - sh - "-c" - | helm template $ARGOCD_APP_NAME -n $ARGOCD_APP_NAMESPACE --include-crds ${ARGOCD_ENV_HELM_ARGS} . | argocd-vault-plugin generate - lockRepo: false avp.yaml: | apiVersion: argoproj.io/v1alpha1 kind: ConfigManagementPlugin metadata: name: argocd-vault-plugin spec: allowConcurrency: true discover: find: command: - sh - "-c" - "find . -name '*.yaml' | xargs -I {} grep \"<path\\|avp\\.kubernetes\\.io\" {} | grep ." generate: command: - argocd-vault-plugin - generate - "." lockRepo: false ---

`apiVersion: apps/v1
kind: Deployment
metadata:
name: argocd-repo-server
spec:
template:
spec:

  # Mount SA token for Kubernets auth
  # Note: In 2.4.0 onward, there is a dedicated SA for repo-server (not default)
  # Note: This is not fully supported for Kubernetes < v1.19
  automountServiceAccountToken: true

  # Each of the embedded YAMLs inside cmp-plugin ConfigMap will be mounted into it's respective plugin sidecar
  volumes:
    - configMap:
        name: cmp-plugin
      name: cmp-plugin
    - name: custom-tools
      emptyDir: {}

  # Download tools
  initContainers:
  - name: download-tools
    image: registry.access.redhat.com/ubi8
    env:
      - name: AVP_VERSION
        value: 1.13.1
    command: [sh, -c]
    args:
      - >-
        curl -L https://github.com/argoproj-labs/argocd-vault-plugin/releases/download/v$(AVP_VERSION)/argocd-vault-plugin_$(AVP_VERSION)_linux_amd64 -o argocd-vault-plugin &&
        chmod +x argocd-vault-plugin &&
        mv argocd-vault-plugin /custom-tools/
    volumeMounts:
      - mountPath: /custom-tools
        name: custom-tools

  # argocd-vault-plugin with Helm
  containers:
  - name: avp-helm
    command: [/var/run/argocd/argocd-cmp-server]
    image: quay.io/argoproj/argocd:v2.4.0
    securityContext:
      runAsNonRoot: true
      runAsUser: 999
    volumeMounts:
      - mountPath: /var/run/argocd
        name: var-files
      - mountPath: /home/argocd/cmp-server/plugins
        name: plugins
      - mountPath: /tmp
        name: tmp

      # Register plugins into sidecar
      - mountPath: /home/argocd/cmp-server/config/plugin.yaml
        subPath: avp-helm.yaml
        name: cmp-plugin

      # Important: Mount tools into $PATH
      - name: custom-tools
        subPath: argocd-vault-plugin
        mountPath: /usr/local/bin/argocd-vault-plugin

  # argocd-vault-plugin with Kustomize
  - name: avp-kustomize
    command: [/var/run/argocd/argocd-cmp-server]
    image: quay.io/argoproj/argocd:v2.4.0
    securityContext:
      runAsNonRoot: true
      runAsUser: 999
    volumeMounts:
      - mountPath: /var/run/argocd
        name: var-files
      - mountPath: /home/argocd/cmp-server/plugins
        name: plugins
      - mountPath: /tmp
        name: tmp

      # Register plugins into sidecar
      - mountPath: /home/argocd/cmp-server/config/plugin.yaml
        subPath: avp-kustomize.yaml
        name: cmp-plugin

      # Important: Mount tools into $PATH
      - name: custom-tools
        subPath: argocd-vault-plugin
        mountPath: /usr/local/bin/argocd-vault-plugin

  # argocd-vault-plugin with plain YAML
  - name: avp
    command: [/var/run/argocd/argocd-cmp-server]
    image: quay.io/argoproj/argocd:v2.4.0
    securityContext:
      runAsNonRoot: true
      runAsUser: 999
    volumeMounts:
      - mountPath: /var/run/argocd
        name: var-files
      - mountPath: /home/argocd/cmp-server/plugins
        name: plugins
      - mountPath: /tmp
        name: tmp

      # Register plugins into sidecar
      - mountPath: /home/argocd/cmp-server/config/plugin.yaml
        subPath: avp.yaml
        name: cmp-plugin

      # Important: Mount tools into $PATH
      - name: custom-tools
        subPath: argocd-vault-plugin
        mountPath: /usr/local/bin/argocd-vault-plugin`
After creating the configMap and patching the argo-repo-server. plugin is working fine when I use to create a kubernetes secret where fetching the secret from vault.

<img width="570" alt="image" src="https://user-images.githubusercontent.com/45903612/219658480-4058599e-bdd6-4c12-b347-9f304e643117.png">


Using the above application and argo-vault-plugin I'm able to fetch the token from vault and create a kubernetes secret.

But when I try deploying the postgresql with helm charts and trying to pass the sensitive values from values.yaml file I'm not able to fetch the secrets from vault directly from values.yaml file. Below are my sample code.

<img width="792" alt="image" src="https://user-images.githubusercontent.com/45903612/219658563-c25dd527-2d1a-462a-a7b1-1b153756705b.png">


<img width="841" alt="image" src="https://user-images.githubusercontent.com/45903612/219658638-7f09b664-4f0b-4460-a8f5-850a9109ccae.png">


Please can you help me how can i achieve this

Describe the solution you'd like
I would like to fetch the secret values from vault directly from values.yaml helm file.

<img width="634" alt="image" src="https://user-images.githubusercontent.com/45903612/219658701-e53c4c72-ca51-4815-ab1d-66a60c92f477.png">


Please let me know if i'm missing something or towards the wrong direction.

**Reply by ciiay** [reply]
Hey @dhinakar2707 , I just started learning vault plugin, and haven't tried to use with helm yet. Have you gone through this blog?
https://luafanti.medium.com/injecting-secrets-from-vault-into-helm-charts-with-argocd-43fc1df57e74

I'm just reading through, and searching for the differences between your example and the example from the blog. It seems to me that you don't have the right environment variables set up for the plugin. 
```
envs:
     - name: VAULT_SECRET_USER
     value: <path:kv-v2/data/demo#user>
     - name: VAULT_SECRET_PASSWORD
     value: <path:kv-v2/data/demo#password>
